### DISCUSSION TITLE
Dimenet trainable parameters

### DISCUSSION THREAD

**Discussion by sudaksh14** [pyg-team/pytorch_geometric#5604] [Q&A]
Hi PyG community,
I was trying to implement the dimenet model from PyG, but while the training my gradients are all nan.
I am not sure as to what parameters should be trained exactly, for now I just used model.parameters().

![dimenet](https://user-images.githubusercontent.com/102596407/193816888-403a5ace-72dd-440c-8c95-8c23ce6a776d.PNG)

**Comment by rusty1s** [comment]
Thanks for reporting. Is it possible to debug this via `with torch.autograd.detect_anomaly()`?

**Reply by sudaksh14** [reply]
Hi @rusty1s , I ran it on the code, it seems that the parameters of the model have no gradient or requires_grad= False.
Could you suggest what should I do?

![image](https://user-images.githubusercontent.com/102596407/193839769-2da01405-bcaf-4bc4-8a1a-5803bf674390.png)

**Reply by rusty1s** [reply]
Interesting. Do you use your own dataset for this? Is it possible to send a minimal example to reproduce?

**Reply by sudaksh14** [reply]
yes, I am using my own dataset for water molecules where I have their xyz co-ordinates.
A code snippet for the same is as follows:
![image](https://user-images.githubusercontent.com/102596407/194058498-607ae1c8-dc32-4dd3-b96e-5ed7415443f6.png)

Then I have converted them into **pos matrix** (3,3) due to 3 atoms only, and I gave an **edge index matrix** as [[1.,0.],[0.,1.]] due to only 2 edges as bonds (I am not sure of this representation, I just thought that this could differentiate b/w the two edges) and finally the **edge attr is same as z input for dimenet forward pass**: torch.tensor([8,1,1]) representing the atomic no of atoms.
Finally the energy is the **target** of my regression task.
I think their maybe some mistake in graph input parameters due to which their is this error but I am not able to identify it, your help would be valuable.
Thnx

**Reply by rusty1s** [reply]
Can you test that using a `LongTensor` for `edge_index`, and using shape `[num_nodes]` for `z`?

**Reply by sudaksh14** [reply]
Hi Matthias,
I tried running with above modifications where edge_index is a LongTensor and z already has dim as [num_nodes] i.e [3] in my case but unfortunately results are still same. Could you identify any issue with datasets maybe as in screenshot below:
PS: I changed the edge_index matrix to 2x3 to add some more sense for the connected nodes as having the value 1 and non-connected as 0. 
![image](https://user-images.githubusercontent.com/102596407/194305008-75964154-37d1-4b8f-9c9f-decafbf4af3f.png)

Thanks a lot for your guidance!!

**Reply by rusty1s** [reply]
I am not 100% certain but there might be an issue in the current `DimeNet` implementation in case of isolates nodes. In your case, you have 3 nodes while only two nodes are connected to each other. What happens if you use something like
```
data.edge_index = torch.tensor([[0, 1, 1, 2], [1, 0, 2, 1]])
```
Can you confirm?

**Reply by sudaksh14** [reply]
Hi Matthias,
I tried the above edge_index, but still the error remains the same as earlier.

So, I was thinking if isolated nodes maybe an issue, then I could try with introducing an artificial edge b/w the isolated nodes maybe and apply their edge attributes as something else say -1 (just an idea), or is it the same thing that you suggested with the above modification?

Thnx
Below are the snaps from your suggested modification:

![image](https://user-images.githubusercontent.com/102596407/195033724-cd485d1c-d8e2-4ef7-958d-1c8cc094dbfb.png)

![image](https://user-images.githubusercontent.com/102596407/195033814-18140dfc-650c-4615-9920-30686d99a74f.png)

**Reply by rusty1s** [reply]
Thanks for reporting! Can you post a minimal example to reproduce this issue? I am happy to look into this.

**Reply by sudaksh14** [reply]
Sure, below code is an example code with only 3 datapoints:

`import torch
from torch import autograd
from torch_geometric.data import Data
from torch_geometric.nn import DimeNet
from tqdm import tqdm
import torch_geometric
from torch_geometric.loader import DataLoader
import numpy as np
import matplotlib.pyplot as plt

edge_index = torch.tensor([[0, 1, 1, 2], [1, 0, 2, 1]])
edge_atr = torch.tensor([8,1,1])

pos = torch.tensor([[[ 0.0089,  0.0000,  0.4355],
         [-0.7973,  0.0000, -0.5977],
         [ 0.6555,  0.0000, -0.4876]],

        [[ 0.0072,  0.0000,  0.4355],
         [-0.7831,  0.0000, -0.5867],
         [ 0.6696,  0.0000, -0.4986]],

        [[-0.0232,  0.0000,  0.4381],
         [-0.5727,  0.0000, -0.4203],
         [ 0.9415,  0.0000, -0.7065]]])

target = torch.tensor([[-76.2732],
        [-76.2739],
        [-76.2507]])


dataset = []
for i in range(3):
    data = Data(pos=pos[i], edge_index=edge_index, edge_attr=edge_atr, y=target[i])
    dataset.append(data)
    
def loss(y_hat,y):
    loss = torch.nn.MSELoss(reduction="mean")
    output = loss(y_hat, y)
    return output

opt = torch.optim.Adam(model.parameters(), lr=1e-5)

def train_on_batch(data, model, opt, loss):
    y_hat = model.forward(z=data.edge_attr, pos=data.pos)
    loss_mean = loss(y_hat, data.y)
    loss_mean.backward()
    opt.step()  
    return loss_mean

model = DimeNet(hidden_channels=128, out_channels=1, num_blocks=6, num_bilinear=8, num_spherical=7, num_radial=6)
model.reset_parameters()

with autograd.detect_anomaly():
    epochs = 10
    train_loss = []
    for epoch in range(epochs):
        for data in tqdm(dataset):
            loss_mean = train_on_batch(data,model,opt,loss)
        print(loss_mean)
        train_loss.append(loss_mean.detach().numpy())

plt.plot(train_loss)

**Reply by EdisonLeeeee** [reply]
I run your code and it works fine on my end. 
```bash
tensor(5798.9111, grad_fn=<MseLossBackward0>)
100%|██████████| 3/3 [00:00<00:00,  8.89it/s]
tensor(5768.4741, grad_fn=<MseLossBackward0>)
100%|██████████| 3/3 [00:00<00:00,  8.91it/s]
tensor(5724.6553, grad_fn=<MseLossBackward0>)
100%|██████████| 3/3 [00:00<00:00,  8.85it/s]
tensor(5662.4111, grad_fn=<MseLossBackward0>)
100%|██████████| 3/3 [00:00<00:00,  8.87it/s]
tensor(5574.9854, grad_fn=<MseLossBackward0>)
100%|██████████| 3/3 [00:00<00:00,  8.83it/s]
tensor(5453.8267, grad_fn=<MseLossBackward0>)
100%|██████████| 3/3 [00:00<00:00,  8.79it/s]
tensor(5287.6616, grad_fn=<MseLossBackward0>)
100%|██████████| 3/3 [00:00<00:00,  8.69it/s]
tensor(5061.3447, grad_fn=<MseLossBackward0>)
100%|██████████| 3/3 [00:00<00:00,  8.82it/s]
tensor(4757.3042, grad_fn=<MseLossBackward0>)
100%|██████████| 3/3 [00:00<00:00,  8.79it/s]
tensor(4356.3213, grad_fn=<MseLossBackward0>)
```
BTW, the optimizer should be defined after the model:
```python
model = DimeNet(hidden_channels=128, out_channels=1, num_blocks=6, num_bilinear=8, num_spherical=7, num_radial=6)
model.reset_parameters()
opt = torch.optim.Adam(model.parameters(), lr=1e-5)
```
I guess that's the problem, right?

**Reply by sudaksh14** [reply]
Yes it works now but only when I use data points separately, but it fails when I use the data loader.  Could you suggest what could be the solution for this?

Thanks a lot!

**Reply by EdisonLeeeee** [reply]
You need to pass `batch` argument to the model when using `DataLoader`:
```python
y_hat = model.forward(z=data.edge_attr, pos=data.pos, batch=data.batch)
```
This should address your problem :)

**Reply by sudaksh14** [reply]
Hi Jintang,
I was already using this in my main code with all the data points.
Below is the link to the repo for the code and data, could you check this one as here I have also defined batches for models but still the gradient is nan.

https://github.com/sudaksh14/Dimenet-GNN.git

Thanks a lot!!

**Reply by EdisonLeeeee** [reply]
Thanks for providing an example to reproduce. I've checked it out and it still works fine on my end by directly running the notebook. I'm using PyG 2.1.0 and PyTorch 1.12+cpu. Will investigate it further and report back.

**Reply by EdisonLeeeee** [reply]
@sudaksh14  Cannot reproduce your bug even with PyTorch 1.11.0. What's your torch_scatter version?

**Reply by sudaksh14** [reply]
Hi Jintang,
below is my pytorch geometric and scatter versions:
![image](https://user-images.githubusercontent.com/102596407/196028455-759166ad-61b6-4ff8-a401-d1d6c720ec61.png)

So in your PC, my code is converging as normal without any nan gradient?  I am not sure why is this not the case in my PC, could this be due to any other reasons like pytorch scatter version?

Thnx a lot!

**Reply by EdisonLeeeee** [reply]
Hi Sudaksh. Everything works fine on my PC. I've checked your notebook again and I found that `data.batch` was not passed to the model. 

Here is the error log from your notebook:
```bash
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Input In [127], in <cell line: 1>()
      5 for data in tqdm(loader_train):
      6     print(data)
----> 7     loss_mean = train_on_batch(data,model,optim1,loss)
      8 print(loss_mean)
      9 train_loss.append(loss_mean.detach().numpy())

Input In [124], in train_on_batch(data, model, opt, loss)
     14 y_hat = model.forward(z=data.edge_attr, pos=data.pos)
     15 loss_mean = loss(y_hat, data.y)
---> 16 loss_mean.backward()
     17 opt.step()  
     18 return loss_mean
```

Could you restart the kernel and rerun the notebook (all the cells) again? This should address your problem.

**Reply by sudaksh14** [reply]
ok I will re-run it, but I have passed data.batch as in below snap:
![image](https://user-images.githubusercontent.com/102596407/196031139-479befc7-138d-4fa5-87bf-eca538416286.png)

The code in the bottom section is just a sample one created with 3 points and hence I didn't give any batches, but the main train_on_batch function I am using is having data.batch.

Thnx.

**Reply by EdisonLeeeee** [reply]
I meant you should run the cell again though you have modified the code correspondingly. Does re-running work for you? Let me know.

**Reply by sudaksh14** [reply]
Yes it is running just now when I used import torch_scatter (maybe it was important), but again there is an issue that when I normalise my data and then run the model it again gives nan gradient (not sure why?)
When I run on data without any scaling its running smoothly but the loss is not converging its like as below:

![image](https://user-images.githubusercontent.com/102596407/196413616-1f6eff4c-452c-4ef2-b78d-5049c6dbf825.png)

Its only for 5 epochs, maybe I can change some hyperparameters for this but don't know why it fails for the scaled data?
Thnx

**Reply by rusty1s** [reply]
How do you normalize your data?

**Reply by sudaksh14** [reply]
I am using the minmax scaler from sklearn and apply that directly to the pd df of my input data and ground truths separately.(2 different scalers)

**Reply by rusty1s** [reply]
Does the minmax scaler produces NaNs if one of the features is all the same? Can you check whether the forward pass already produces NaN values?

**Reply by EdisonLeeeee** [reply]
I don't think it is a necessary way to normalize the data in such a regression task. I've run the code with the minmax normalization, and the output of the model is close to zero while the ground truth `y` is much larger than the output (e.g., 76). Hence the gradient would become large and the loss comes to NaNs after a few iterations of gradient descent.

**Reply by sudaksh14** [reply]
In my data all the y coordinate is 0 due to molecules in same plane, hence it could be the issue I think.
I checked with **np.argwhere(np.isnan(nodes))** that minmax scaler doesn't produce any nan values. Still I tried to just scale the x&z coordinates only(leaving y-coordinate) and the target values, then tried to train the network. It now gives me very high loss value (like in e+15).
So do you suggest dropping the scaling altogether as its leading to such high losses? Or is there any other way around

Thanks a lot! Really Appreciate your time and effort.

**Reply by EdisonLeeeee** [reply]
As the problem is related to the very high loss, or gradients in other words. Could you have a try on gradient clipping? For example:
```python
loss.backward()
torch.nn.utils.clip_grad_norm_(model.parameters(), grad_norm) # grad_norm could be 1.0 or 2.0, it depends.
```
